PurchasesCoreSwift/Public/PurchaserInfo.swift

```diff
@@ -0,0 +1,326 @@
+
+            if let originalPurchaseDateString = subscriberData["original_purchase_date"] as? String {
+                self.originalPurchaseDate = PurchaserInfo.parseDate(
+                    dateString: originalPurchaseDateString,
+                    withDateFormatter: PurchaserInfo.dateFormatter)
+            } else {
+                self.originalPurchaseDate = nil
+            }
+
+            guard let firstSeenDateString = subscriberData["first_seen"] as? String,
+                  let firstSeenDate = PurchaserInfo.parseDate(
+                    dateString: firstSeenDateString,
+                    withDateFormatter: dateFormatter) else {
+                return nil
+            }
+            self.firstSeen = firstSeenDate
+
+            guard let originalAppUserIdString = subscriberData["original_app_user_id"] as? String else {
+                return nil
+            }
+            self.originalAppUserId = originalAppUserIdString
+
+            if let managementUrlString = subscriberData["management_url"] as? String {
+                self.managementURL = parseURL(url: managementUrlString)
+            } else {
+                self.managementURL = nil
+            }
+
+            // Purchases and entitlements
+            self.nonSubscriptions =
+                subscriberData["non_subscriptions"] as? [String: [[String: Any]]] ?? [String: [[String: Any]]]()
+            self.entitlements = subscriberData["entitlements"] as? [String: Any] ?? [String: Any]()
+            self.nonSubscriptionTransactions = TransactionsFactory().nonSubscriptionTransactions(
+                withSubscriptionsData: nonSubscriptions,
+                dateFormatter: dateFormatter)
+
+            let latestNonSubscriptionTransactionsByProductId =
+                [String: Any](uniqueKeysWithValues: nonSubscriptions.map { productId, transactionsArray in
+                (productId, transactionsArray.last)
+            })
+
+            self.allPurchases =
+                latestNonSubscriptionTransactionsByProductId.merging(subscriptions) { (current, _) in current }
+
+            self.expirationDatesByProduct = PurchaserInfo.parseExpirationDate(expirationDates: subscriptions)
+            self.purchaseDatesByProduct = PurchaserInfo.parsePurchaseDate(purchaseDates: allPurchases)
+        }
+    }
+
+}
+
+private extension PurchaserInfo {
+
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set(dates.keys.filter {
+            guard let nonNullDate = dates[$0] as? Date else { return true }
+            return isAfterReferenceDate(date: nonNullDate)
+        })
+    }
+
+    func isAfterReferenceDate(date: Date) -> Bool { date.timeIntervalSince(self.requestDate) > 0 }
+
+    class func parseURL(url: String) -> URL? { URL(string: url) }
+
+    class func parseDate(dateString: String, withDateFormatter: DateFormatter) -> Date? {
```

Comment:
the `withDataFormatter` style isn't really needed for swift, this could be 
```swift
class func parseDate(dateString: String, dateFormatter: DateFormatter) -> Date? {
```

Reply:
```swift
let firstSeenDate = PurchaserInfo.parseDate(dateString: originalPurchaseDateString,
                                            withDateFormatter: PurchaserInfo.dateFormatter)
// vs
let firstSeenDate = dateFormatter.date(from: originalPurchaseDateString)
```

Reply:
I'm now realizing that these are probably static because we have a static date formatter.
And formatters are costly to build, so we really only want to have one, so that's why we have it as static. 

I think the best solution for a situation like this one is to have the formatter be injected into the `init`, and have a singleton instance living somewhere else. 

I believe you had it like that and I made you change it 🤦 That's entirely my bad. I think what had seemed odd to me was the syntax `DateFormatter.dateFormatter.foo`. 

However, if we had something like 
- `DateFormatter.iso8601Formatter` as a static property in an extension to `DateFormatter`, ideally outside this file
- had a property in PurchaserInfo `private let dateFormatter: DateFormatter`
- send it in the `init`, i.e.: 
```swift
init?(subscriberData: [String: Any], dateFormatter: DateFormatter = .iso8601DateFormatter()) {
    self.dateFormatter = dateFormatter
    ...
```
then we can use it as 

```swift
let firstSeenDate = dateFormatter.date(from: originalPurchaseDateString)
```

and, should we ever need to test that the formatting is done appropriately, we can always mock the formatter here, and we even have quick access to writing tests for the `iso8601Formatter` to make sure that we're creating it correctly.

Reply:
after writing this it dawned on me that we already have a date formatter built in, for the local receipt parser, so we can reuse [that one](https://github.com/RevenueCat/purchases-ios/blob/main/PurchasesCoreSwift/LocalReceiptParsing/DataConverters/ISO3601DateFormatter.swift)
I wish I'd thought to name it `iso8601DateFormatter` back then, though.